### ISSUE TITLE
Support fontconfig and/or specifying multiple custom font dirs

### ISSUE THREAD

**Issue by GoldsteinE** [typst/typst#100]
Some usecases (e.g. packaging on NixOS) require specifying multiple custom font location, which is currently impossible. The best way to do it would be IMO support fontconfig. The second best way is to support `$XDG_DATA_DIRS` environment variable instead of only supporting `$XDG_DATA_HOME`.

Probably related to #97.

**Comment by 7sDream** [comment]
I just made a PR for this, it adds custom font search path support. 

And, btw, I think [fontdb](https://github.com/RazrFalcon/fontdb) is a good crate to replace the font searcher in cli, it's from the same author of ttf-parser, and [has basic support for fontconfig](https://github.com/RazrFalcon/fontdb/blob/c63d68c9d148598ee119fd04fcf6368363b1375d/src/lib.rs#L383). It doesn't use fontconfig c lib, but parse it's configure file instead.

**Comment by drupol** [comment]
On my side, I have the variable `FONTCONFIG_FILE` set with custom fonts, but `typst --fonts` doesn't return anything.

**Comment by 7sDream** [comment]
@drupol fontdb will [works with this env var](https://github.com/RazrFalcon/fontdb/blob/c63d68c9d148598ee119fd04fcf6368363b1375d/src/lib.rs#L389).

But I think I should discuss with the author(and other developers) whether this solution is acceptable, before working on a new PR to land it.

**Comment by laurmaedje** [comment]
Interesting that fontdb supports this now. Last time I checked it did pretty much what Typst does now.

**Comment by laurmaedje** [comment]
Using fontdb seems reasonable to me, in principle. Haven't looked at the details though.

**Comment by 7sDream** [comment]
looks worth trying, maybe we can merge #270 first, and then I can continue work on fontdb intergation from here?

**Comment by KiruyaMomochi** [comment]
Had a look at fontdb. It levearges crate [fontconfig_parser](https://crates.io/crates/fontconfig-parser) to [parse](https://github.com/RazrFalcon/fontdb/blob/c63d68c9d148598ee119fd04fcf6368363b1375d/src/lib.rs#L386-L398) `fonts.conf`.
We can also use this parser crate directly.

## fontdb + fontconfig_parser

In fontdb, location of `fonts.conf` is given by `FONTCONFIG_FILE`, or by merging these files:
- `$XDG_CONFIG_HOME/fontconfig/fonts.conf`
- `/etc/fonts/local.conf`
- `/etc/fonts/fonts.conf`.

I'm not sure why `local.conf` is searched, according to [fonts-conf(5)](https://www.freedesktop.org/software/fontconfig/fontconfig-user.html) this is included by `fonts.conf` already.

## fontconfig-rs + libfontconfig

Another option is using FFI to call the real fontconfig library. While the best option is using fontconfig to match fonts, it's easier to just get font directories.

There are binding crates like [fontconfig-rs](https://github.com/yeslogic/fontconfig-rs), but this crate doesn't expose `FcConfigGetFontDirs` so we must do it our own.

## Experimentation

I made an experiment to show the difference between them, you can find code and result at https://github.com/KiruyaMomochi/fontconfig-lab

It seems that their results are almost same, although
- the order may differ, and
- `libfontconfig` also considers fonts under `$XDG_DATA_DIRS/share/fonts`. However, none of them exists on my OS.

**Comment by 7sDream** [comment]
@KiruyaMomochi 

Nice work! I prefer to parse config file too, as Typst is pure Rust for now and I just don't want to break it. 😆 

BTW, I originally planned to start integrating fontdb today, but because you have already written some code, and I am a Windows user, it is not very convenient for me to test the result, so maybe it is more appropriate for you to take over the follow-up intergation work? So I can have a rest on sunday haha 🍹 

What do you think?

**Comment by KiruyaMomochi** [comment]
Yes, although I'm not good at typesetting, I can give it a try.

But there is a problem. And decision needs to be made.

fontdb **[directly loads font](https://github.com/RazrFalcon/fontdb/blob/c63d68c9d148598ee119fd04fcf6368363b1375d/src/lib.rs#L225)** into [`fontdb::FaceInfo`](https://github.com/RazrFalcon/fontdb/blob/c63d68c9d148598ee119fd04fcf6368363b1375d/src/lib.rs#L669). Comparing to our [`typst::font::book::FontInfo`](https://github.com/typst/typst/blob/e70ec5f3c06312b7ff2388630e05e3c2d745896f/src/font/book.rs#L148), **their `FaceInfo` does not have [`coverage`](https://github.com/typst/typst/blob/e70ec5f3c06312b7ff2388630e05e3c2d745896f/src/font/book.rs#L157) field**.

That means if we want to *use fontdb*, and at the same time use *FontInfo*, we may need to **read each font twice**.

Some workarounds can be
1. fork their fontdb, then a) add a coverage to them or, b) replace their `FaceInfo` with our `FontInfo`.
2. use fontconfig_parser crate directly.

**Comment by 7sDream** [comment]
Yes, but the doubled call is just `ttf-parse::RawFace::parse`, which should be pretty cheap, IIRC.

**Comment by 7sDream** [comment]
@KiruyaMomochi @drupol @GoldsteinE 

jsut did some quick coding: https://github.com/7sDream/typst/compare/main..cli/fontdb

It works just like before in my computer, but can you test it to see if the fontconfig thing works for you?

**Comment by KiruyaMomochi** [comment]
It works for me too.

<details>

<summary>Screenshot</summary>

![Screenshot_20230326_174421](https://user-images.githubusercontent.com/65301509/227767648-aebec97f-9c44-41ca-99f8-c1eee929dcfb.png)

</details>

**Comment by 7sDream** [comment]
@KiruyaMomochi And custom font path in your config file automaticlly works too?

**Comment by 7sDream** [comment]
As RazrFalcon/fontdb#50 has been merged, I think I can open a new PR after fontdb bumps a new version.

And then we can get a simplest work version, i.e., you have a fontconfig file in the default location `~/.config/fontconfig/fonts.conf` and your custom font dir is directlly in that file.

**Comment by Andrew15-5** [comment]
I was having a problem using my fonts. They, as most downloadable fonts, are in `~/.fonts/`. But from reading the comments, it is not the directory that is read by Typst (although LuaLaTeX does a great job of finding all my fonts). I tried adding a relative symlink to `~/.local/share/fonts/` (the dir didn't even exist) and now it found all my fonts.

**Comment by AlexD97** [comment]
Does font detection work now on NixOS? For me, it does not work.

**Comment by 7sDream** [comment]
@AlexD97 This is a very old fork, but would still love to hear your feedback.

Recently fontdb has released some new versions and my PR is landed in version 0.13.1.

However Riey/fontconfig-parser#3 by @KiruyaMomochi has been merged but not released yet. We need this to solve a bug: "include" in fontconfig configuration files didn't work.

**Comment by AlexD97** [comment]
As I've already commented, more fonts are recognized with your fork. More specifically, the fonts that are installed “directly” through nix are recognized. The fonts that are managed by home-manager are not detected (I have the `fonts.fontconfig.enable` option set to `true`).

**Comment by 7sDream** [comment]
@AlexD97 Although I often hear about NixOS, I haven't used it in depth, and I don't know much about some of its concepts. I think we can make the fontconfig configuration works in "normal" cases. If this is not enough for NixOS, we can add support for it on top of this.

BTW, fontconfig-parser has bumped a new version 0.5.3, so I think the fontdb integration work can be started. I can do this but my day job will be busy until september, so anyone can take over it if interested :)

**Comment by laurmaedje** [comment]
Nice, that's a surprisingly simple change. And the existing code was adapted from `fontdb` anyway.

**Comment by AlexD97** [comment]
@7sDream Thanks! The fork works nicely on my machine (now, even fonts that are installed through home-manager are recognized).

**Comment by drupol** [comment]
Can't wait for this to be comitted!
